### ISSUE TITLE
Constructing replacement text for fixes without bleed or side effects

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#7836]
see :
- https://github.com/stylelint/stylelint/pull/7802#discussion_r1653241078
- https://github.com/stylelint/stylelint/issues/7784

The purpose of this issue is to explore how we can construct replacement texts for fixes at a more theoretical level.

This issue assumes that we will require rules to expose replacement text and ranges to solve : https://github.com/stylelint/stylelint/issues/7192

### Multiple parsers / AST's

We don't have a single AST, we use multiple parsers and produce multiple distinct AST's.

This implies that some rules work in multiple stages:
1. parse a string with a specialized parser
2. detect a problem in the specialized AST
3. warn about the problem or apply a fix
4. serialize the specialized AST
5. apply the modified string to the PostCSS AST

Multiple subsequent fixes will mutate the same AST and serializing the entire AST will contain previous fixes.

This is bad because an external tool should be able to apply fixes individually.

This is the example given in : https://github.com/stylelint/stylelint/pull/7802#discussion_r1653241078

### Constructing replacement text is non-trivial

We prefer to work with parsed values from an AST when detecting and fixing problems.
The replacement text however is what is present in the output.

One is decoded and unescaped, the other is encoded and might contain escape sequences.

With the `@csstools/css-tokenizer` we store and expose both:

```css
/**
 * The CSS Token interface
 *
 * @remarks
 * CSS Tokens are fully typed and have a strict structure.
 * This makes it easier to iterate and analyze a token stream.
 *
 * The string representation and the parsed value are stored separately for many token types.
 * It is always assumed that the string representation will be used when stringifying, while the parsed value should be used when analyzing tokens.
 */
export interface Token<T extends TokenType, U> extends Array<T | string | number | U> {
	/**
	 * The type of token
	 */
	0: T,
	/**
	 * The token representation
	 *
	 * @remarks
	 * This field will be used when stringifying the token.
	 * Any stored value is assumed to be valid CSS.
	 *
	 * You should never use this field when analyzing the token when there is a parsed value available.
	 * But you must store mutated values here.
	 */
	1: string,
	/**
	 * Start position of representation
	 */
	2: number,
	/**
	 * End position of representation
	 */
	3: number,
	/**
	 * Extra data
	 *
	 * @remarks
	 * This holds the parsed value of each token.
	 * These values are unescaped, unquoted, converted to numbers, etc.
	 *
	 * You should always use this field when analyzing the token.
	 * But you must not assume that mutating only this field will have any effect.
	 */
	4: U,
}
```

In other words, the fix values for an AST mutation are different from the replacement text as requested in : https://github.com/stylelint/stylelint/issues/7192

We absolute can't take the values from current fixers and return those as the replacement text.

One way around this problem is to mutate AST's and to serialize them, as we can assume that specialized parsers can handle mutations and serialization following the CSS specs.
But this loops back to the first point, that multiple fixes on the same AST affect each other.

Another way around this is to construct new AST's, or to clone existing AST's.
This gives the benefit of correct serialization but without introducing the bleed effects.

### Performance

The `detection`, `warn` and `fix` execution paths have very different characteristics.

- `detection` always runs and should be as fast and lightweight as possible
- `warn` runs often, but not always. It should still be fast if possible
- `fix` only runs once and then the problem is gone. It is fine if this is a bit more intensive (time and memory)

With the planned changes to expose replacement text, the `warn` path will also include most of the work of the `fix` path.

If we return replacement text as part of the `warn` path then we will slow down all warnings because more work will happen.

We should be careful to pick the right patterns here so that we minimize the impact.

## Ideal solution

_This doesn't exist, and I don't think it is realistic, but I keep coming back to it._

To me this feels like database transactions.

You want to start a transaction, do some reads and writes and then optionally commit or rollback the changes.

It would be super handy if AST's had transactional mutations.

**Comment by romainmenke** [comment]
🤔 Most of the problems outlined above can be solved naively by ignoring performance.

We can construct and clone AST's.
We can serialize with specialized parsers to ensure that escaping and encoding is correct.

These things are not impossible, they are just an extra step.
This implies more code to write in rules but also slower execution.

If we make the construction of `EditInfo` (`range` and `replacement`) an optional thing we can reduce the performance impact for must cases.

Most users and contexts won't need `EditInfo`.

I think that this information is useful for editor integrations as they can then have tooltips for individual fixes.
Or when users set a specific flag/option.

Any problem that will be fixed also doesn't need to generate `EditInfo`, you never need both.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for sharing your thoughts and explaining the problems.

I realize finding an ideal API to support both autofix and `EditInfo` construction is challenging. As you pointed out, such an editor info support is a nice-to-have feature. We should not sacrifice performance for that purpose.

So, until we find a great solution, providing an experiment API might be an option. For example, it might be necessary to be separated from the `fix` callback like this:

```js
report({
	fix: () => {
		// Mutate PostCSS AST objects...
	},

	editInfo: () => {
		// Calculate a range and find a replacement text...
		return { range, replacement };
	},
});
```

If doing so, plugin authors would have to calculate them by themselves. In my honest opinion, implementing it seems a pain to many people... 🤔

**Comment by romainmenke** [comment]
I think I might have found a way forwards that is relatively simple and gives good results.

See: https://github.com/stylelint/stylelint/pull/8234

The central idea is to return a PostCSS node that encloses all mutations.
And to use data already available in PostCSS to calculate what was changed and where.

This gives rule authors a really simple API and doesn't require multiple callbacks (i.e. `fix()`, `editInfo()`)

There are some compromises:
- it is impossible to add support for overlapping fixes (eslint also has this limitation)
- it is impossible to have both this feature enabled and the regular `fix` feature

I think this is fine.

-----

If this direction seems interesting and worthwhile to everyone (@pamelalozano16, @ybiquitous, ...) then I will outline some next steps.

There are still some edge cases where I am unsure how everything will behave.

**Comment by Mouvedia** [comment]
> And to use data already available in PostCSS to calculate what was changed and where.

In https://github.com/stylelint/stylelint/issues/8233#issuecomment-2563108932, I was guessing either events or diffing; you went with the latter it seems.

> it is impossible to have both this feature enabled and the _regular_ `fix` feature

Can you develop a bit? Is this akin to `context.fix` which has been deprecated?
Looking at #8234, it doesn't seem you dropped the `fix` callback.
i.e. only its return type changed

**Comment by romainmenke** [comment]
> > it is impossible to have both this feature enabled and the regular fix feature

> Can you develop a bit?

For this feature we make the assumption that no output CSS will ever be consumed by the user.
We assume that when a user sets `computeReplacementText` to true that they do not want anything to be altered in their source files.

_Possibly we need to throw when a user sets `{ fix: true, computeReplacementText: true }`, but that is beside the point. Better to punt such a detail for later._

We make these assumptions because of the use case for `computeReplacementText`.
The intended use is to gather data to feed into another tool. As this other cool will apply the actual fixes. Such a pipeline is non-sensical if Stylelint also applies fixes to source code. This would make the gathered data invalid. (the reported ranges for example would be incorrect)

Users either want to fix immediately or gather data to use later, never both.

----

We can use that knowledge to assume that it is safe to mutate the AST when `computeReplacementText` is true as the output will never be used anyway.

This allows us to simply run the existing `fix()` callback.
In turn this implies less changes and code to write for rule authors.

----

I don't see a way to have both featured enabled at the same time and have meaningful and useful outputs.

Disallowing such combinations makes it easier to know how this will be used and what we can assume to be safe.

**Comment by romainmenke** [comment]
> Looking at https://github.com/stylelint/stylelint/pull/8234, it doesn't seem you dropped the fix callback.
> i.e. only its return type changed

Not sure why we would drop the `fix()` callback :)
We spend a great deal of time designing that feature and it works really well.

Yes the options (in general) that were under consideration were:
- add arguments to the `fix()` callback
- add return values to the `fix()` callback
- add more fields to `Problem`
- add another callback to `Problem`

In all cases we were focussed on the replacement text and the range of the original text.
We never considered using any other kind of value.
Using a `PostCSS node` as the return value for `fix()` is new in this latest proposal.

Technically it could also be a field on `Problem` :

```js
{
  fix: () => { ... },
  fixedNode: someNode
}
```

It doesn't need to be the return value of the callback and I don't really have any strong opinions on what is a better fit.

**Comment by Mouvedia** [comment]
> Possibly we need to throw when a user sets { fix: true, computeReplacementText: true }, but that is beside the point.

That helped me a lot. I was just guessing until then.
It is not OR nor AND it's XOR.
It's clearer now.

I suppose that it will be accompanied later on by a `disableComputeReplacementText` in the same way that have a `disableFix` for granularity.

> Users either want to fix immediately or gather data to use later, never both.

OK.

> I don't see a way to have both featured enabled at the same time and have meaningful and useful outputs.

If a rule has `disableFix: false,` I would expect `computeReplacementText: true` to be applied to it.

**Comment by romainmenke** [comment]
> I suppose that it will be accompanied later on by a `disableComputeReplacementText` in the same way that have a [disableFix](https://stylelint.io/user-guide/configure/#disablefix) for granularity.

Possibly 🤔 
It really depends on how this will be used.

This extra data is part of `Warning` and users can already manage warnings in various ways.
But we should hear about this from end users.

> If a rule has `disableFix: false` I would expect `computeReplacementText: true` to not be ignored.

It really is all or nothing :)
Either you are fixing or you are computing replacement text.

You can't mix and match.
You can't fix by some rules and compute the replacement text by some other rules.

**Comment by Mouvedia** [comment]
> It doesn't need to be the return value of the callback and I don't really have any strong opinions on what is a better fit.

Let's discuss that in #7784 or the PR directly once you are ready for a review.
The better solution would probably be the one that:
1. doesn't perform stuff unnecessarily
2. require the less effort for plugin authors to refactor
3. is future-proof

> It really is all or nothing.

Alright, K.I.S.S. then.

**Comment by romainmenke** [comment]
Closing as discussed in https://github.com/stylelint/stylelint/issues/7192#issuecomment-2568284585
